components/Carousel/Carousel.jsx

```diff
@@ -0,0 +1,188 @@
+      width: 0, // passed down to <Slide> so it can reuse the h/w calculations
+    };
+    this.el = null;
+    this.setProportionalHeight = this.setProportionalHeight.bind(this);
+    this.keyboardNavigate = this.keyboardNavigate.bind(this);
+    this.generateCoin = this.generateCoin.bind(this);
+    this.goToSlide = this.goToSlide.bind(this);
+    this.next = this.next.bind(this);
+    this.prev = this.prev.bind(this);
+  }
+
+  componentDidMount() {
+    this.setProportionalHeight();
+    window.addEventListener('resize', this.setProportionalHeight);
+    // NOTE: if keyboard navigation ends up being an issue because of <input> elements on the page,
+    // maybe bind the event to `this.el` instead of `window`.
+    window.addEventListener('keyup', this.keyboardNavigate);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.setProportionalHeight);
+    window.removeEventListener('keyup', this.keyboardNavigate);
+  }
+
+  setProportionalHeight() {
+    if (this.el) {
+      const MOBILE_PADDING_BOTTOM = 0;
+      const { maxHeight, minHeight } = this.props;
+      const width = this.el.clientWidth;
+      const aspectHeight = getAspectRatioHeight(this.props.aspectRatio, width);
+      const height = containValue(maxHeight, minHeight, aspectHeight);
+      const isMobile = height > getAspectRatioHeight('16:9', width);
+      this.setState({ height, isMobile, width });
+      this.el.style.height = `${height + (isMobile ? MOBILE_PADDING_BOTTOM : 0)}px`;
+    }
+  }
+
+  keyboardNavigate(event) {
+    const isAnimating = this.state.exitDirection !== '';
+    if (isAnimating || this.props.slides.length <= 1) { return; }
+    const [ LEFT, RIGHT ] = [ 37, 39 ];
+    const key = event.keyCode || event.which;
+    if (key === LEFT) { this.prev(); }
+    if (key === RIGHT) { this.next(); }
+  }
+
+  goToSlide(i, overrideDirection = '') {
+    const direction = i > this.state.topSlideIndex ? (overrideDirection || 'TO_LEFT') : (overrideDirection || 'TO_RIGHT');
+    this.setState({
+      bgSlideIndex: i,
+      enterDirection: direction,
+      exitDirection: direction,
+      isFresh: false,
+    });
+
+    setTimeout(() => {
+      this.setState({
+        bgSlideIndex: i, // This is very odd. bgSlideIndex *should* be assumed to be the result of calcNext(). In some cases, that leads to wrong animations. But for some reason this works.
+        enterDirection: this.state.exitDirection,
+        exitDirection: '',
+        topSlideIndex: this.state.bgSlideIndex, // === i
+      });
+    }, this.props.animationDuration);
+  }
+
+  next() {
+    const nextSlide = calcNext(this.props.slides.length, this.state.topSlideIndex);
+    this.goToSlide(nextSlide, 'TO_LEFT');
+  }
+
+  prev() {
+    const prevSlide = calcPrev(this.props.slides.length, this.state.topSlideIndex);
+    this.goToSlide(prevSlide, 'TO_RIGHT');
+  }
+
+  generateCoin(Slide, i) {
+    const isAnimating = this.state.exitDirection !== '';
```

Comment:
Since this check is happening in 3 different places, it might be good to pull this out into it's own function